src/data_race.rs

```diff
@@ -0,0 +1,1369 @@
+    /// this should always mirror the local value in thread.rs
+    /// the thread name is used for improved diagnostics
+    /// during a data-race.
+    #[inline]
+    pub fn thread_set_name(&self, thread: ThreadId, name: String) {
+        let name = name.into_boxed_str();
+        let mut thread_info = self.thread_info.borrow_mut();
+        thread_info[thread].thread_name = Some(name);
+    }
+
+    /// Attempt to perform a synchronized operation, this
+    /// will perform no operation if multi-threading is
+    /// not currently enabled.
+    /// Otherwise it will increment the clock for the current
+    /// vector before and after the operation for data-race
+    /// detection between any happens-before edges the
+    /// operation may create.
+    fn maybe_perform_sync_operation<'tcx>(
+        &self,
+        op: impl FnOnce(VectorIdx, RefMut<'_, ThreadClockSet>) -> InterpResult<'tcx>,
+    ) -> InterpResult<'tcx> {
+        if self.multi_threaded.get() {
+            let (index, mut clocks) = self.current_thread_state_mut();
+            clocks.increment_clock(index);
+            op(index, clocks)?;
+            let (_, mut clocks) = self.current_thread_state_mut();
+            clocks.increment_clock(index);
+        }
+        Ok(())
+    }
+
+    /// Internal utility to identify a thread stored internally
+    /// returns the id and the name for better diagnostics.
+    fn print_thread_metadata(&self, vector: VectorIdx) -> String {
+        let thread = self.vector_info.borrow()[vector];
+        let thread_name = &self.thread_info.borrow()[thread].thread_name;
+        if let Some(name) = thread_name {
+            let name: &str = name;
+            format!("Thread(id = {:?}, name = {:?})", thread.to_u32(), &*name)
+        } else {
+            format!("Thread(id = {:?})", thread.to_u32())
+        }
+    }
+
+    /// Acquire a lock, express that the previous call of
+    /// `validate_lock_release` must happen before this.
+    pub fn validate_lock_acquire(&self, lock: &VClock, thread: ThreadId) {
+        let (index, mut clocks) = self.load_thread_state_mut(thread);
+        clocks.increment_clock(index);
+        clocks.clock.join(&lock);
+        clocks.increment_clock(index);
+    }
+
+    /// Release a lock handle, express that this happens-before
+    /// any subsequent calls to `validate_lock_acquire`.
+    pub fn validate_lock_release(&self, lock: &mut VClock, thread: ThreadId) {
+        let (index, mut clocks) = self.load_thread_state_mut(thread);
+        clocks.increment_clock(index);
+        lock.clone_from(&clocks.clock);
+        clocks.increment_clock(index);
+    }
+
+    /// Release a lock handle, express that this happens-before
+    /// any subsequent calls to `validate_lock_acquire` as well
+    /// as any previous calls to this function after any
+    /// `validate_lock_release` calls.
+    pub fn validate_lock_release_shared(&self, lock: &mut VClock, thread: ThreadId) {
```

Comment:
So, the difference between "validate_lock_release" and "validate_lock_release_shared" is that the former *overwrites* the old clock, but the latter joins them.

For a normal lock, shouldn't it be the case that both are equivalent? Since the thread that released the lock previously acquired it, its clock will be ahead of the lock clock, so `join` and `clone_from` do the same? (I am trying to figure out why both of these functions exist.)

Also, I noticed "validate_lock_release" is used for condvars, not just for locks. That makes the naming a bit confusing, though I do not have a better proposal. Is this use for condvars related to using `clone_from` vs `join`?